### PR TITLE
Add memory supervision to script execution supervisor

### PR DESCRIPTION
**Pull request by janakamarasena**
This PR provides the capability to limit the amount of memory that can be consumed by an adaptive auth script when it is being executed.

By default, no memory limits are imposed. If a memory limit needs to be set add the following config to the deployment.toml file

memory_limit is set in bytes.
```
[authentication.adaptive.execution_supervisor]
memory_limit = "500000"
```

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java [hunk 1]
```diff
@@ -82,13 +116,12 @@ public void monitor(String identifier, String serviceProvider, String tenantDoma
      * @param identifier Monitoring task identifier.
      * @return Total elapsed time of adaptive auth execution in the current thread.
      */
-    public long completed(String identifier) {
+    public JSExecutionMonitorResult completed(String identifier) {
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:119]: Don't we want to stop monitoring for memory?

Something like
`memoryCounter.setThreadAllocatedMemoryEnabled(false);`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java [hunk 2]
```diff
@@ -128,21 +165,48 @@ public MonitoringTask getMonitoringTask() {
...
             this.id = id;
             this.serviceProvider = serviceProvider;
             this.tenantDomain = tenantDomain;
             this.timeCreated = System.currentTimeMillis();
             this.elapsedTimeInMillis = elapsedTimeInMillis;
+            this.consumedMemoryInBytes = consumedMemoryInBytes;
+
+            if (memoryLimitInBytes > 0) {
+                java.lang.management.ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+                if (threadMXBean instanceof com.sun.management.ThreadMXBean) {
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:196]: not needed, but added it for better readability coz the previous line also has a ThreadMXBean

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java [hunk 3]
```diff
@@ -384,8 +384,22 @@ private void setAdaptiveAuthExecutionSupervisor() {
...
 
+        String memoryLimitString = IdentityUtil.getProperty(
+                FrameworkConstants.AdaptiveAuthentication.CONF_EXECUTION_SUPERVISOR_MEMORY_LIMIT);
+        long memoryLimitInBytes = FrameworkConstants.AdaptiveAuthentication.DEFAULT_EXECUTION_SUPERVISOR_MEMORY_LIMIT;
+        if (StringUtils.isNotBlank(memoryLimitString)) {
+            try {
+                memoryLimitInBytes = Long.parseLong(memoryLimitString);
+            } catch (NumberFormatException e) {
+                log.error("Error while parsing adaptive authentication execution supervisor memory limit config: "
+                        + memoryLimitString + ", setting memory limit to default value: " + memoryLimitInBytes, e);
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/FrameworkServiceComponent.java]: changed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 4]
```diff
@@ -961,21 +960,42 @@ private JSExecutionSupervisor getJSExecutionSupervisor() {
...
-                FrameworkConstants.AdaptiveAuthentication.PROP_EXECUTION_SUPERVISOR_ELAPSED_TIME);
-        if (elapsedTimeObj != null) {
-            elapsedTime = (Long) elapsedTimeObj;
+        JSExecutionMonitorResult jsExecutionMonitorResult;
+        Object storedResult = context.getProperty(
+                FrameworkConstants.AdaptiveAuthentication.PROP_EXECUTION_SUPERVISOR_RESULT);
+        if (storedResult != null) {
+            jsExecutionMonitorResult = (JSExecutionMonitorResult) storedResult;
+        } else {
+            jsExecutionMonitorResult = new JSExecutionMonitorResult(0, 0);
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java]: changed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 5]
```diff
@@ -961,21 +960,42 @@ private JSExecutionSupervisor getJSExecutionSupervisor() {
         return FrameworkServiceDataHolder.getInstance().getJsExecutionSupervisor();
     }
 
-    private void setAuthScriptExecutionElapsedTime(AuthenticationContext context, long elapsedTime) {
+    private void setAuthScriptExecutionMonitorResult(AuthenticationContext context,
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java]: changed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java (hunk 1 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java]: Changed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionMonitorResult.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.config.model.graph;
+
+import java.io.Serializable;
+
+/**
+ * Holds the result of a script execution monitor.
+ */
+public class JSExecutionMonitorResult implements Serializable {
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionMonitorResult.java]: changed

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionMonitorResult.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionMonitorResult.java]: fixed

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**